### DISCUSSION TITLE
No customiser in the visualiser

### DISCUSSION THREAD

**Discussion by lf94** [sgenoud/replicad#38] [Visualiser]
For some reason it's just not appearing for me...? Try this:

```js
const defaultParams = {
  outerRingDiameter: 44.0,
  innerRingDiameter: 38.0,
  topThickness: 3.0,
  gripDiameter: 2.0,
  socketDepth: 10.0,
  gripsTotal: 6,
};

const main = ({ sketchCircle, Sketcher, Plane }, {
  outerRingDiameter,
  innerRingDiameter,
  topThickness,
  gripDiameter,
  socketDepth,
  gripsTotal,
}) => {
  const top  = sketchCircle(outerRingDiameter / 2).extrude(topThickness);
  const base = sketchCircle((innerRingDiameter - (gripDiameter / 2)) / 2)
    .loftWith(sketchCircle((innerRingDiameter - (gripDiameter / 2) - 1.0) / 2, new Plane([0, 0, -socketDepth])));
  const baseWithTop = top.fuse(base);
  const withCut = baseWithTop.cut(
    sketchCircle(
      innerRingDiameter / 2 - 2.0,
      new Plane([0, 0, -socketDepth])
    )
    .extrude(socketDepth + topThickness)
  );
  const withTopFillet1 = withCut.fillet(topThickness - 0.01, e => e.containsPoint([0, outerRingDiameter / 2, topThickness]));
  const withTopFillet2 = withTopFillet1.fillet(topThickness / 2, e => e.containsPoint([0, innerRingDiameter / 2 - 2, topThickness]));

  const socket = withTopFillet2;

  const parentRadii = { start: (innerRingDiameter - gripDiameter / 2) / 2, end: ((innerRingDiameter - gripDiameter / 2) - 1.0) / 2 };
  const childRadii = { start: gripDiameter / 2, end: gripDiameter / 2 };

  const grips = getAngles(gripsTotal).map(angle =>
    modelGrip({ sketchCircle, Plane })({ angle, height: socketDepth, parentRadii, childRadii })
  );
  return grips.reduce((acc, s) => acc.fuse(s), socket);
};

const toDegrees = radians => (radians / (Math.PI * 2)) * 360;
const toRadians = angle => (angle / 360) * (Math.PI * 2);

const modelGrip = ({ sketchCircle, Plane }) => ({ angle, height, parentRadii, childRadii }) => {
  const radiusDiff = parentRadii.start - parentRadii.end;

  const parallelAngle = (90 - toDegrees(Math.atan(height / radiusDiff)));
  const grip = sketchCircle(childRadii.start)
  .loftWith(sketchCircle(childRadii.end, new Plane([0, 0, -height])))
  .rotate(parallelAngle, [0,0,0], [Math.cos(toRadians(angle + 90)),Math.sin(toRadians(angle + 90)),0])
  .translate([
    parentRadii.start * Math.cos(toRadians(angle)),
    parentRadii.start * Math.sin(toRadians(angle)),
    0
  ]);
  return grip;
};

const getAngles = pieSlices => new Array(pieSlices).fill(0).map((_, index) => (360 / pieSlices) * index);

```

**Comment by lf94** [comment]
@sgenoud any ideas?

**Comment by raydeleu** [comment]
I tested this with the bird house example that is on the Replicad.xyz site. Indeed the customizer is not appearing, so I guess the error is not in your code. Strangely enough the customizer does appear if you select the examples directly in the Visualiser.  My assumption would be that if you select the example a direct link to the website is opened with a different visualiser. See the images below, the interface looks different. 

![comparison-visualiser](https://user-images.githubusercontent.com/38007983/176303588-978dda57-d596-4e58-a7f7-6025b7cd0a45.png)

**Comment by sgenoud** [comment]
The customiser does not appear in the visualiser, only on shared models (or in the studio gallery). I want to refactor a bit the tools of the visualiser and this should let met have a way to include the customiser. I would like to add the following improvements:

* allow more configuration when exporting STL
* include the customiser
* give some information on faces / edges selected
* make it possible to show a slide the model to see the inside
* see the console log within the main screen


I have made this a discussion - I would like to keep the issues for the library only, and put the rest of the ecosystem in the discussion!

**Comment by sgenoud** [comment]
I have done some work to be able to improve significantly the visualiser. You can play with a first iteration of the new visualiser - with support for `defaultParams`

**Reply by raydeleu** [reply]
<img width="908" alt="customizer" src="https://user-images.githubusercontent.com/38007983/177218584-a8efe357-4258-49a2-ad73-1a25ea0e571e.png">

Works nice! However there seems to be an issue updating the local web app. I had to delete the visualiser from `users/test/Applications/Chrome Apps/` and re-install the new version to make this work. Without this update the app kept asking to enter the file name. If you use it directly from `replicad.xyz` there is no problem.
